polygon/heimdall/simulator/bor_events.go

```diff
@@ -0,0 +1,45 @@
+package simulator
+
+import (
+	"context"
+	"time"
+
+	"github.com/ledgerwatch/erigon-lib/downloader/snaptype"
+	"github.com/ledgerwatch/erigon/polygon/heimdall"
+	"github.com/ledgerwatch/erigon/turbo/snapshotsync/freezeblocks"
+)
+
+func continueLoop(events []*heimdall.EventRecordWithTime, to time.Time, limit int) bool {
+	if len(events) == 0 {
+		return true
+	}
+	if len(events) == limit {
+		return false
+	}
+
+	last := events[len(events)-1]
+	return last.Time.Before(to)
+}
+
+func (h *HeimdallSimulator) getEvents(ctx context.Context, fromId uint64, to time.Time, limit int) ([]*heimdall.EventRecordWithTime, error) {
+	span, err := h.blockReader.EventsById(fromId, to, limit)
+
+	view := h.knownBorSnapshots.View()
+	defer view.Close()
+
+	for continueLoop(span, to, limit) && err == nil {
+		if seg, ok := view.EventsSegment(h.lastDownloadedBlockNumber); ok {
+			if err := h.downloadData(ctx, seg, snaptype.BorEvents, freezeblocks.BorEventsIdx); err != nil {
+				return nil, err
+			}
+		}
+		h.lastDownloadedBlockNumber += 500000
+
+		span, err = h.blockReader.EventsById(fromId, to, limit)
```

Comment:
suggest to rename "span" to "events" and do an if err != nil { return err } check